### DISCUSSION TITLE
Update including daemon and gossipsub discovery

### DISCUSSION THREAD

**Discussion by notassigned** [alecbcs/hyprspace#47] [Show and tell]
I've made quite a few changes under the hood and I would like to get some feedback. The code can be cleaned up a bit but the idea was to make it more reliable and quick.

Changes include:
* Faster LUT for packet switching using ipv4 addrs as uint32 instead of string
* Gossipsub for peer discovery
* Daemon which can be started in foreground with `sudo hyprspace daemon up`
* Command to get currently connected peers `hyprspace peers <interface name>`
* Configurable transports (quic, tcp)
* Configurable auto relay (on/off)

https://github.com/notassigned/hyprspace/tree/reliability-connectivity-upgrade

**Comment by alecbcs** [comment]
@notassigned this looks good! 
> - Daemon which can be started in foreground with `sudo hyprspace daemon up`

Any particular reason you changed to this instead of the existing?
```
hyprspace up --foreground hs0
```

Also have you done any performance testing of these changes? Both network throughput and reconnect times?

I'd say go ahead and submit this as a PR when you are ready and then we can work on flushing out some of this together.

**Reply by notassigned** [reply]
So I added back in the --foreground flag on the up command, but the reason that I removed it is because you would only be able to use it on one interface. e.g starting on the first interface would work, but the second foreground interface will just print 'daemon is already running' or something. I know we could fix this by capturing output of the daemon and streaming it to the program, but that's gonna require some effort. Another possible solution is to just start the interface on the current program, not by calling the daemon at all.

On performance: I tested it against the current build and the background cpu usage is comparable, but under load my branch performs much better. Opening and closing streams for each IP packet adds most of the overhead as well as 1-2ms of latency, and converting each IP packet header's dest field to a string adds some more. Under load my branch uses only slightly more cpu than when idling.

Discovery is faster as well, and more efficient. This can be improved further by having peers share other peers they are connected to in the network, but for now it just uses gossipsub.

Running the read loops uses virtually no cpu, even without the 1ms sleep that I added to reduce cycles when the data length is 0. I guess goroutines are pretty efficient :)

**Reply by alecbcs** [reply]
> So I added back in the --foreground flag on the up command, but the reason that I removed it is because you would only be able to use it on one interface. e.g starting on the first interface would work, but the second foreground interface will just print 'daemon is already running' or something. I know we could fix this by capturing output of the daemon and streaming it to the program, but that's gonna require some effort. Another possible solution is to just start the interface on the current program, not by calling the daemon at all.

Ahh gotcha! So if you don't mind me asking why the new daemon system? I'm often running multiple hyprspace daemon instances in the background so it might be problematic if you can only run one at a time now.

Nice on the performance testing! Do you have a way to measure bandwidth throughput with something like iperf?

**Reply by notassigned** [reply]
> Ahh gotcha! So if you don't mind me asking why the new daemon system? I'm often running multiple hyprspace daemon instances in the background so it might be problematic if you can only run one at a time now.

To clarify, you can still run multiple instances, just not in the foreground. I added it so that we could do things that require IPC like the peers command which prints the connected peers on an interface. It also lets you cleanly shut down interfaces by calling the shutdown function rather than just deleting the tun interface. If you want multiple interfaces running, they all run under the daemon process, you just don't start them in the foreground, but if you want to see the daemon output for some reason you can start it in the foreground. We can also add other commands that print the status/statistics/history now that we have IPC.

> Nice on the performance testing! Do you have a way to measure bandwidth throughput with something like iperf?

I haven't used that but I'll check it out.

**Reply by notassigned** [reply]
I also forgot to mention that now that separate interfaces are being run by the same program, further optimization could be done to reduce cpu usage and connection time like sharing a DHT between all interfaces.

**Reply by alecbcs** [reply]
@notassigned Oh interesting! By having multiple interfaces run on the same program does that mean that all of the interfaces share a single internal libp2p node?

**Reply by notassigned** [reply]
> Oh interesting! By having multiple interfaces run on the same program does that mean that all of the interfaces share a single internal libp2p node?

They don't right now, each interface still uses it's own full libp2p node, but what we could do is have the daemon start a full libp2p node with an ephemeral identity which bootstraps the local dht and can join gossipsub topics on behalf of an interface with a bare bones node for discovery which would allow hundreds or thousands of hyprspace interfaces to be run with the same resource usage as a single current one.

Also, I tested my build with iperf and the bandwidth sucks (about 25% that of the main branch) so I ran wireshark and found that there is constant TCP re-transmission going on. I'm still trying to debug why that's happening.

**Reply by notassigned** [reply]
Some packets are out of order.... I am losing my mind.

This guy has the same issue it seems. If you know how to fix this, lmk.
https://stackoverflow.com/questions/67779021/golang-tun-udp-vpn-very-slow-tcp-connections-with-lots-of-out-of-order-segment

**Reply by alecbcs** [reply]
Hey @notassigned, out of order packets shouldn't matter? It's common for packets to be delivered out of order on the internet and that's why TCP has built in systems for tracking packet order and re-requesting lost packets. So if any application is using TCP on top of Hyprspace that TCP connection should take care of reassembling the packets not us. Also if any application is using UDP there are no order guarantees and thus the application itself should be able to deal with different packet orders.

If you're getting network slowdowns that might be because packets are actually being lost/taking so long/so out of order that the system is re-requesting them multiple times.

**Reply by notassigned** [reply]
Yeah, it just seemed a bit excessive but I guess it's ok. I'll open a PR.
